clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java

```diff
@@ -561,6 +561,9 @@ private void producePutOrDeleteToKafka(
           beforeProcessingRecordTimestamp);
     } else {
       int valueLen = updatedValueBytes.remaining();
+
+      // Make sure there is capacity in the drainer queue before occupying space in the transient record cache
+      maybeBlockOnDrainerCapacity();
       partitionConsumptionState.setTransientRecord(
```

Comment:
Is it ideal to precede every call to `PartitionConsumptionState::setTransientRecord` with this new function? Would it make sense to include this functionality inside of the `setTransientRecord` function itself? This may require passing a closure to evaluate the buffer service's remaining capacity into the PCS, which is maybe not ideal from an encapsulation standpoint. On the other hand, having this convention that calls to `setTransientRecord` should always be preceded by `maybeBlockOnDrainerCapacity` seems a bit fragile as well. Maybe there should be a function in `LFSIT` to wrap both of these functions?

Reply:
Yeah, I think we should do the closure.  Will add.

Reply:
Gonna add this here because for some reason I can't reply directly to the sleep commnet???

> would be great to avoid sleeping, if possible

Any recommendation on a different pattern?  Gaojie seemed to prefer sleep to throwing an exception.